File: Chapter10/src/main/java/chapter10/EnumExample.java
package chapter10;

/**
 * Created by alexeysoshin on 13/04/2018.
 */
public class EnumExample {
    public static void main(String[] args) {
        PizzaOrderStatus status = PizzaOrderStatus.ORDER_RECEIVED;
        System.out.println(status.nextStatus().nextStatus());
    }
}


enum PizzaOrderStatus {
    ORDER_RECEIVED,
    PIZZA_BEING_MADE,
    OUT_FOR_DELIVERY,
    COMPLETED;


    public PizzaOrderStatus nextStatus() {
        switch (this) {
            case ORDER_RECEIVED: return PIZZA_BEING_MADE;
            case PIZZA_BEING_MADE: return OUT_FOR_DELIVERY;
            case OUT_FOR_DELIVERY: return COMPLETED;
            case COMPLETED: return COMPLETED;
        }
        throw new RuntimeException("Unhandled status " + this);
    }
}

File: Chapter10/src/main/kotlin/chapter10/EnumExample.kt
package chapter10

import chapter10.PizzaOrderStatus.*
fun main(vararg args: String) {

    var status: PizzaOrderStatus = OrderReceived(123)

    while (status !is Completed) {
        status = when (status) {
            is OrderReceived -> status.nextStatus()
            is PizzaBeingMade -> status.nextStatus()
            is OutForDelivery -> status.nextStatus()
            is Completed -> status
        }
    }
}

sealed class PizzaOrderStatus(protected val orderId: Int) {
    abstract fun nextStatus() : PizzaOrderStatus
    class OrderReceived(orderId: Int) : PizzaOrderStatus(orderId) {
        override fun nextStatus(): PizzaOrderStatus {
            return PizzaBeingMade(orderId)
        }
    }

    class PizzaBeingMade(orderId: Int) : PizzaOrderStatus(orderId) {
        override fun nextStatus(): PizzaOrderStatus {
            return OutForDelivery(orderId)
        }
    }

    class OutForDelivery(orderId: Int) : PizzaOrderStatus(orderId) {
        override fun nextStatus(): PizzaOrderStatus {
            return Completed(orderId)
        }
    }

    class Completed(orderId: Int) : PizzaOrderStatus(orderId) {
        override fun nextStatus(): PizzaOrderStatus {
            return this
        }
    }
}


